tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,223 @@
+from zarr.store import MemoryStore
+
+# from strategies_store import StoreStatefulStrategies, key_ranges
+from zarr.testing.strategies import key_ranges, paths
+
+# zarr spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html
+
+
+class SyncStoreWrapper:
+    def __init__(self, store):
+        """Class to hold sync functions that map to async methods of MemoryStore
+        MemoryStore methods are async, this class' methods are sync, so just need to call asyncio.run() in them
+        then, methods in statemachine class are sync and call sync.
+        Unfortunately, hypothesis' stateful testing infra does not support asyncio
+        So we redefine sync versions of the Store API.
+        https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+        """
+        self.store = store
+        self.mode = store.mode
+
+    def set(self, key, data_buffer):  # buffer is value
+        return asyncio.run(self.store.set(key, data_buffer))
+
+    def list(self):
+        async def wrapper(gen):
+            return [i async for i in gen]
+
+        gen = self.store.list()  # async store list
+        return (i for i in asyncio.run(wrapper(gen)))
+
+    def get(self, key):
+        obs = asyncio.run(self.store.get(key, prototype=default_buffer_prototype()))
+        return obs
+
+    def get_partial_values(self, key_ranges):
+        obs_partial = asyncio.run(
+            self.store.get_partial_values(
+                prototype=default_buffer_prototype(), key_ranges=key_ranges
+            )
+        )
+        return obs_partial
+
+    def delete(self, path):  # path is key
+        return asyncio.run(self.store.delete(path))
+
+    def empty(self):
+        return asyncio.run(self.store.empty())
+
+    def clear(self):
+        return asyncio.run(self.store.clear())
+
+    def exists(self, key):
+        return asyncio.run(self.store.exists(key))
+
+    def list_dir(self, prefix):
+        raise NotImplementedError
+
+    def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    def supports_listing(self):
+        raise NotImplementedError
+
+    def supports_partial_writes(self):
+        raise NotImplementedError
+
+    def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    keys_bundle = Bundle("keys_bundle")
+    key_ranges_bundle = Bundle("key_ranges")
+
+    def __init__(self):
+        super().__init__()
+        self.model: dict[str, bytes] = {}
+        self.store = SyncStoreWrapper(MemoryStore(mode="w"))
+
+    # NOTE: set this and the next as initialize instead of rule to make sure they run
+    #      and create key, key_ranges before any rules that use key, key_ranges
+    #      not totally sure this is necessary
+    @initialize(key=paths, data=st.binary(min_size=0, max_size=100), target=keys_bundle)
```

Comment:
```suggestion
    @rule(key=paths, data=st.binary(min_size=0, max_size=100), target=keys_bundle)
```

yes should not be necessary.

`initialize` only runs once so this is not good for exploring all possibilities.